sql: write epoch based table descriptor leases

vivekmenezes opened the pull request.

vivekmenezes committed:
sqlbase: add epoch to the table descriptor lease table

related to #23978

Release note: None

vivekmenezes committed:
sql: write epoch based table descriptor leases

This change begins to write epoch based table descriptor leases
after a cluster upgrade. The expiration time for the lease is
still set so these leases still act as expiration based leases.
The deadline for transactions using the lease is still the
expiration time of the lease. This change is focused on the
renewal process of upgrading these leases correctly when the
node liveness epoch changes.

This change is to be followed by another change (part 2) that will
implement the withdrawal of such leases by other nodes in the cluster
when the epoch for the originating node changes. Note we expect to
use longer epoch liveness intervals for these leases. When
part 2 is implemented we can start setting the txn deadline based on
the liveness expiration time of the lease (which is a sliding
deadline).

related to #23978

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/28508)
<!-- Reviewable:end -->

bdarnell reviewed (changes requested):
I can't evaluate this PR on its own without seeing how you plan to use this epoch information. Andrei and I raised questions about the soundness of this scheme in #28458 and those questions haven't been answered. Let's continue the discussion in that issue. 

> Reviewed 3 of 3 files at r1, 10 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28508#-:-LJp6swxAcE8uajkNHIQ:b-ysxafb)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/base/config.go, line 94 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28508#-LJp3vx28UPOjLXrNLHN:-LJp3vx28UPOjLXrNLHO:b-t0u8mk) ([raw file](https://github.com/cockroachdb/cockroach/blob/f718271feba44be694f76487f13b0577e2dfff22/pkg/base/config.go#L94)):*
> ```Go
> 	// DefaultTableDescriptorLeaseEpochRefreshInterval is the default
> 	// periodicity at which an attempt is made to refresh the LeaseManager
> 	// local epoch.
> ```

What does this mean? What would happen if this value were higher or lower? 
___
*[pkg/sql/lease.go, line 1811 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28508#-LJp6SlJ9TmCkurWWek8:-LJp6SlJ9TmCkurWWek9:b-ekody8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f718271feba44be694f76487f13b0577e2dfff22/pkg/sql/lease.go#L1811)):*
> ```Go
> 		m.mu.Unlock()
> 		defer close(done)
> 		// Acquire leases in waves.
> ```

It would make more sense to limit the size of a "wave" than the number of them. It would be even better to use a semaphore to run continuously at a given concurrency level instead of batching things arbitrarily. 
___
*[pkg/sqlmigrations/migrations.go, line 851 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28508#-LJp23Yq4M4DBe_Qzs5K:-LJp23Yq4M4DBe_Qzs5L:b-llld83) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc87682fae6b094542defa72d316067c2ed65644/pkg/sqlmigrations/migrations.go#L851)):*
> ```Go
> }
> 
> func addEpochToTableDescriptorLeaseTable(ctx context.Context, r runner) error {
> ```

This could add more than just the epoch if we make another change to the descriptor in the same release cycle. If this goes into 2.1 there's not much risk of that, but there is risk of this pattern being copied. Why isn't this using the pattern of `addJobsProgress` (to take the most immediate example)?  Or should we give the `node` user permission to alter these system tables? 


<!-- Sent from Reviewable.io -->

vivekmenezes reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28508)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/base/config.go, line 94 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28508#-LJp3vx28UPOjLXrNLHN:-LJsOMnPC92d6wZY-lKN:bw1b1hp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f718271feba44be694f76487f13b0577e2dfff22/pkg/base/config.go#L94)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

What does this mean? What would happen if this value were higher or lower? 
</blockquote></details>

Added some more comments to explain the drawbacks of setting this to a high value.
___
*[pkg/sql/lease.go, line 1811 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28508#-LJp6SlJ9TmCkurWWek8:-LJsLv8mBfjpTLUnEy3q:b-oztqnb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f718271feba44be694f76487f13b0577e2dfff22/pkg/sql/lease.go#L1811)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It would make more sense to limit the size of a "wave" than the number of them. It would be even better to use a semaphore to run continuously at a given concurrency level instead of batching things arbitrarily. 
</blockquote></details>

Changed to use RunLimitAsyncTask() instead. 
___
*[pkg/sqlmigrations/migrations.go, line 851 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28508#-LJp23Yq4M4DBe_Qzs5K:-LJsMFtZ1mvu2yYLp-yP:b-jb1fhj) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc87682fae6b094542defa72d316067c2ed65644/pkg/sqlmigrations/migrations.go#L851)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This could add more than just the epoch if we make another change to the descriptor in the same release cycle. If this goes into 2.1 there's not much risk of that, but there is risk of this pattern being copied. Why isn't this using the pattern of `addJobsProgress` (to take the most immediate example)?  Or should we give the `node` user permission to alter these system tables? 
</blockquote></details>

I'm not happy with the addJobProgress style of adding a schema change because it could create a table descriptor that is not the exact same as the one present on new clusters. I prefer not changing the security model for the system tables.

Should I change this to changeTableDescriptorLeaseTable() so that any future change in this release will get added here?


<!-- Sent from Reviewable.io -->

vivekmenezes committed:
sql: change LeaseManager API to export lease epoch

The lease epoch will eventually be used to look up
an expiration time for the epoch.

Update TestAllLeasesRenewedWhenEpochChanges to check
for epoch returned.

ajwerner commented:
Epoch-based schema leases are not something we want to move forward with at this point.
